semgrep/semgrep/core_runner.py

```diff
@@ -561,6 +561,7 @@ def _run_rules_direct_to_semgrep_core(
         rule_file_name = (
             RULE_SAVE_FILE
             if dump_command_for_core
+            # we dump JSON formatted rules to .yaml because core's JSON parser can't parse them
```

Comment:
what is the semgrep-core error?

Reply:
I think semgrep-core can parse rules as .yaml, or .json (and also .jsonnet)

Reply:
@aryx letting core parse as JSON causes a pattern parse error. My guess would be that backslash escaping in strings is not parsed correctly by the JSON parser.

Reply:
Do you have an example showing the issue. We actuallu use @mjambon json parser

Reply:
I think if you download p/default and dump it as JSON, it'll just fail and you can debug from there

Reply:
@underyx can you please post the json data that causes the parsing error?

Reply:
[rules.json.zip](https://github.com/returntocorp/semgrep/files/8733776/rules.json.zip)

This is it:

```
❯ semgrep-core -rules /tmp/rules.json
/tmp/rules.json:205:21: "Pattern parse error": Invalid pattern for Python
```

Reply:
That line is

```
"pattern": "itsdangerous.jws.JSONWebSignatureSerializer(..., algorithm_name=\"none\", ...)"
```

which is why I assumed that core doesn't parse `\"` within strings correctly.

Reply:
thanks, @underyx.

@pad are we not using the yaml parser for this, not a json-only parser? It's very suspicious that it would not parse this correctly, though. This is line 205:
```
          "pattern": "itsdangerous.jws.JSONWebSignatureSerializer(..., algorithm_name=\"none\", ...)"
```

Reply:
Here's a smaller example with the same error:
```
{
  "rules": [
    {
      "id": "insecure-itsdangerous-use",
      "message": "The Python third-party 'itsdangerous' module used with 'none' signing algorithm",
      "languages": [
        "python"
      ],
      "severity": "WARNING",
      "metadata": {
        "source_rule_url": "https://github.com/dlint-py/dlint/blob/master/docs/linters/DUO137.md",
        "category": "security",
        "technology": [
          "itsdangerous"
        ],
        "license": "Commons Clause License Condition v1.0[LGPL-2.1-only]"
      },
      "pattern-either": [
        {
          "pattern": "itsdangerous.jws.JSONWebSignatureSerializer(..., algorithm_name=\"none\", ...)"
        },
        {
          "pattern": "itsdangerous.JSONWebSignatureSerializer(..., algorithm_name=\"none\", ...)"
        }
      ]
    }
  ]
}
```

```
semgrep-core -rules rules.json
rules.json:20:21: "Pattern parse error": Invalid pattern for Python
```

Reply:
This works:
```
{
  "rules": [
    {
      "id": "x",
      "message": "",
      "languages": ["python"],
      "severity": "WARNING",
      "pattern": "'hello'"
    }
  ]
}
```
->
```
$ semgrep-core -lang python -rules rules.json
```

The problem is with double-quoted strings:
```
{
  "rules": [
    {
      "id": "x",
      "message": "",
      "languages": ["python"],
      "severity": "WARNING",
      "pattern": "\"hello\""
    }
  ]
}
```
->
```
$ semgrep-core -lang python -rules rules.json
rules.json:8:17: "Pattern parse error": Invalid pattern for Python
```

```

Reply:
Two findings:
1. OCaml's yaml package parses `"\"hello\""` correctly.
2. Passing the pattern directly to semgrep-core with `-e` also works as expected.

1:
```
utop # Yaml.of_string {|{
  "rules": [
    {
      "id": "x",
      "message": "",
      "languages": ["python"],
      "severity": "WARNING",
      "pattern": "\"hello\""
    }
  ]
}|};;
- : Yaml.value Yaml.res =
Ok
 (`O
    [("rules",
      `A
        [`O
           [("id", `String "x"); ("message", `String "");
            ("languages", `A [`String "python"]);
            ("severity", `String "WARNING"); ("pattern", `String "\"hello\"")]])])

```

2:
```
$ semgrep-core -lang python -e '"hello"' <(echo '"hello"')
/tmp/semgrep-core-552fe4-63:1
 "hello"
```

I haven't looked at the semgrep-core source code yet.

Reply:
Same error with more logging:
```
$ semgrep-core -lang python -rules rules.json
rules.json:8:17: "Pattern parse error": Invalid pattern for Python:
--- pattern ---
\"hello\"
--- end pattern ---
Pattern error: Parse_info.Lexical_error (unrecognized symbol: \, /tmp/tmp-2268658-8712a8.py:1:0 "\\")
```

Reply:
I located the bug, in the custom yaml parser where the string literal should be evaluated and isn't. I'll file a proper bug report.

Reply:
I created https://github.com/returntocorp/semgrep/issues/5268 and https://github.com/returntocorp/semgrep/issues/5269

Reply:
BTW @underyx when you see those kinds of bugs in semgrep-core, it's better to tell us than writing some comments in the code (I happened to review this diff by chance, but otherwise I would have missed the bug we have in semgrep-core).

Reply:
@aryx I was gonna, don't worry

Reply:
Got it. Btw i've tried to fix it so maybe we can retry this json thing at some point.